In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
# 1.1.1 Split data into train and test data
from sklearn.model_selection import train_test_split

# 1.1.2 PReprocessong:
from sklearn.preprocessing import StandardScaler

# 1.1.3 Import class DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier as dtree

# 1.1.4 To draw decision tree
from sklearn.tree import export_graphviz
from sklearn.tree import plot_tree
import matplotlib.pyplot as plt
from sklearn.tree import export_graphviz
 
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from sklearn.metrics import classification_report
from sklearn.datasets import load_iris
import os

from sklearn.model_selection import train_test_split

from sklearn.tree import plot_tree
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('Income Evaluation.csv')

In [3]:
df.head()

age          workclass   fnlwgt   education   education-num  \
0   39          State-gov    77516   Bachelors              13   
1   50   Self-emp-not-inc    83311   Bachelors              13   
2   38            Private   215646     HS-grad               9   
3   53            Private   234721        11th               7   
4   28            Private   338409   Bachelors              13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

    capital-gain   capital-loss   hours-per-week  native-country  income  
0           2174              0               40   United-States   <=50K  
1              0              0               13   United-States   <=50K  
2              0              0               40   United-States   <=50K  
3              0              0               40   United-States   <=50K  
4              0              0               40            Cuba   <=50K

In [4]:
df.describe()

age        fnlwgt   education-num   capital-gain  \
count  32561.000000  3.256100e+04    32561.000000   32561.000000   
mean      38.581647  1.897784e+05       10.080679    1077.648844   
std       13.640433  1.055500e+05        2.572720    7385.292085   
min       17.000000  1.228500e+04        1.000000       0.000000   
25%       28.000000  1.178270e+05        9.000000       0.000000   
50%       37.000000  1.783560e+05       10.000000       0.000000   
75%       48.000000  2.370510e+05       12.000000       0.000000   
max       90.000000  1.484705e+06       16.000000   99999.000000   

        capital-loss   hours-per-week  
count   32561.000000     32561.000000  
mean       87.303830        40.437456  
std       402.960219        12.347429  
min         0.000000         1.000000  
25%         0.000000        40.000000  
50%         0.000000        40.000000  
75%         0.000000        45.000000  
max      4356.000000        99.000000

In [5]:
data = df.values
print(data)

[[39 ' State-gov' 77516 ... 40 ' United-States' ' <=50K']
 [50 ' Self-emp-not-inc' 83311 ... 13 ' United-States' ' <=50K']
 [38 ' Private' 215646 ... 40 ' United-States' ' <=50K']
 ...
 [58 ' Private' 151910 ... 40 ' United-States' ' <=50K']
 [22 ' Private' 201490 ... 20 ' United-States' ' <=50K']
 [52 ' Self-emp-inc' 287927 ... 40 ' United-States' ' >50K']]


In [6]:
df.isna().sum()

age                0
 workclass         0
 fnlwgt            0
 education         0
 education-num     0
 marital-status    0
 occupation        0
 relationship      0
 race              0
 sex               0
 capital-gain      0
 capital-loss      0
 hours-per-week    0
 native-country    0
 income            0
dtype: int64

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              32561 non-null  int64 
 1    workclass       32561 non-null  object
 2    fnlwgt          32561 non-null  int64 
 3    education       32561 non-null  object
 4    education-num   32561 non-null  int64 
 5    marital-status  32561 non-null  object
 6    occupation      32561 non-null  object
 7    relationship    32561 non-null  object
 8    race            32561 non-null  object
 9    sex             32561 non-null  object
 10   capital-gain    32561 non-null  int64 
 11   capital-loss    32561 non-null  int64 
 12   hours-per-week  32561 non-null  int64 
 13   native-country  32561 non-null  object
 14   income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [8]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [9]:
df[' workclass'] = le.fit_transform(df[' workclass'])
df[' education'] = le.fit_transform(df[' education'])
df[' marital-status'] = le.fit_transform(df[' marital-status'])
df[' occupation'] = le.fit_transform(df[' occupation'])
df[' relationship'] = le.fit_transform(df[' relationship'])
df[' race'] = le.fit_transform(df[' race'])
df[' sex'] = le.fit_transform(df[' sex'])
df[' native-country'] = le.fit_transform(df[' native-country'])
df[' income'] = le.fit_transform(df[' income'])

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   age              32561 non-null  int64
 1    workclass       32561 non-null  int32
 2    fnlwgt          32561 non-null  int64
 3    education       32561 non-null  int32
 4    education-num   32561 non-null  int64
 5    marital-status  32561 non-null  int32
 6    occupation      32561 non-null  int32
 7    relationship    32561 non-null  int32
 8    race            32561 non-null  int32
 9    sex             32561 non-null  int32
 10   capital-gain    32561 non-null  int64
 11   capital-loss    32561 non-null  int64
 12   hours-per-week  32561 non-null  int64
 13   native-country  32561 non-null  int32
 14   income          32561 non-null  int32
dtypes: int32(9), int64(6)
memory usage: 2.6 MB


In [11]:
df.corr()

age   workclass    fnlwgt   education   education-num  \
age              1.000000    0.003787 -0.076646   -0.010508        0.036527   
 workclass       0.003787    1.000000 -0.016656    0.023513        0.052085   
 fnlwgt         -0.076646   -0.016656  1.000000   -0.028145       -0.043195   
 education      -0.010508    0.023513 -0.028145    1.000000        0.359153   
 education-num   0.036527    0.052085 -0.043195    0.359153        1.000000   
 marital-status -0.266288   -0.064731  0.028153   -0.038407       -0.069304   
 occupation     -0.020947    0.254892  0.001597   -0.021260        0.109697   
 relationship   -0.263698   -0.090461  0.008931   -0.010876       -0.094153   
 race            0.028718    0.049742 -0.021291    0.014131        0.031838   
 sex             0.088832    0.095981  0.026858   -0.027356        0.012280   
 capital-gain    0.077674    0.033835  0.000432    0.030046        0.122630   
 capital-loss    0.057775    0.012216 -0.010252    0.016746        0.079923   
 hours-per-week  0.068756    0.138962 -0.018768    0.055510        0.148123   
 native-country -0.001151   -0.007690 -0.051966    0.064288        0.050840   
 income          0.234037    0.051604 -0.009463    0.079317        0.335154   

                  marital-status   occupation   relationship      race  \
age                    -0.266288    -0.020947      -0.263698  0.028718   
 workclass             -0.064731     0.254892      -0.090461  0.049742   
 fnlwgt                 0.028153     0.001597       0.008931 -0.021291   
 education             -0.038407    -0.021260      -0.010876  0.014131   
 education-num         -0.069304     0.109697      -0.094153  0.031838   
 marital-status         1.000000    -0.009654       0.185451 -0.068013   
 occupation            -0.009654     1.000000      -0.075607  0.006763   
 relationship           0.185451    -0.075607       1.000000 -0.116055   
 race                  -0.068013     0.006763      -0.116055  1.000000   
 sex                   -0.129314     0.080296      -0.582454  0.087204   
 capital-gain          -0.043393     0.025505      -0.057919  0.011145   
 capital-loss          -0.034187     0.017987      -0.061062  0.018899   
 hours-per-week        -0.190519     0.080383      -0.248974  0.041910   
 native-country        -0.023819    -0.012543      -0.005507  0.137852   
 income                -0.199307     0.075468      -0.250918  0.071846   

                      sex   capital-gain   capital-loss   hours-per-week  \
age              0.088832       0.077674       0.057775         0.068756   
 workclass       0.095981       0.033835       0.012216         0.138962   
 fnlwgt          0.026858       0.000432      -0.010252        -0.018768   
 education      -0.027356       0.030046       0.016746         0.055510   
 education-num   0.012280       0.122630       0.079923         0.148123   
 marital-status -0.129314      -0.043393      -0.034187        -0.190519   
 occupation      0.080296       0.025505       0.017987         0.080383   
 relationship   -0.582454      -0.057919      -0.061062        -0.248974   
 race            0.087204       0.011145       0.018899         0.041910   
 sex             1.000000       0.048480       0.045567         0.229309   
 capital-gain    0.048480       1.000000      -0.031615         0.078409   
 capital-loss    0.045567      -0.031615       1.000000         0.054256   
 hours-per-week  0.229309       0.078409       0.054256         1.000000   
 native-country -0.008119      -0.001982       0.000419        -0.002671   
 income          0.215980       0.223329       0.150526         0.229689   

                  native-country    income  
age                    -0.001151  0.234037  
 workclass             -0.007690  0.051604  
 fnlwgt                -0.051966 -0.009463  
 education              0.064288  0.079317  
 education-num          0.050840  0.335154  
 marital-status        -0.023819 -0.199307  
 occupation            -0.012543  0.0754

In [12]:
#dropping the fnlwgt column because of its negligible correlation with income
df.drop(' fnlwgt', axis=1, inplace= True)

In [13]:
df.head()

age   workclass   education   education-num   marital-status   occupation  \
0   39           7           9              13                4            1   
1   50           6           9              13                2            4   
2   38           4          11               9                0            6   
3   53           4           1               7                2            6   
4   28           4           9              13                2           10   

    relationship   race   sex   capital-gain   capital-loss   hours-per-week  \
0              1      4     1           2174              0               40   
1              0      4     1              0              0               13   
2              1      4     1              0              0               40   
3              0      2     1              0              0               40   
4              5      2     0              0              0               40   

    native-country   income  
0               39        0  
1               39        0  
2               39        0  
3               39        0  
4                5        0

In [14]:
x = df.iloc[:,0:13]
y = df.iloc[:,13]

In [15]:
x

age   workclass   education   education-num   marital-status  \
0       39           7           9              13                4   
1       50           6           9              13                2   
2       38           4          11               9                0   
3       53           4           1               7                2   
4       28           4           9              13                2   
...    ...         ...         ...             ...              ...   
32556   27           4           7              12                2   
32557   40           4          11               9                2   
32558   58           4          11               9                6   
32559   22           4          11               9                4   
32560   52           5          11               9                2   

        occupation   relationship   race   sex   capital-gain   capital-loss  \
0                1              1      4     1           2174              0   
1                4              0      4     1              0              0   
2                6              1      4     1              0              0   
3                6              0      2     1              0              0   
4               10              5      2     0              0              0   
...            ...            ...    ...   ...            ...            ...   
32556           13              5      4     0              0              0   
32557            7              0      4     1              0              0   
32558            1              4      4     0              0              0   
32559            1              3      4     1              0              0   
32560            4              5      4     0          15024              0   

        hours-per-week   native-country  
0                   40               39  
1                   13               39  
2                   40               39  
3                   40               39  
4                   40                5  
...                ...              ...  
32556               38               39  
32557               40               39  
32558               40               39  
32559               20               39  
32560               40               39  

[32561 rows x 13 columns]

In [16]:
y

0        0
1        0
2        0
3        0
4        0
        ..
32556    0
32557    1
32558    0
32559    0
32560    1
Name:  income, Length: 32561, dtype: int32

In [17]:
from sklearn.preprocessing import StandardScaler
ss= StandardScaler()
ss.fit(x)
x = ss.transform(x)

In [18]:
x

array([[ 0.03067056,  2.15057856, -0.33543693, ..., -0.21665953,
        -0.03542945,  0.29156857],
       [ 0.83710898,  1.46373585, -0.33543693, ..., -0.21665953,
        -2.22215312,  0.29156857],
       [-0.04264203,  0.09005041,  0.18133163, ..., -0.21665953,
        -0.03542945,  0.29156857],
       ...,
       [ 1.42360965,  0.09005041,  0.18133163, ..., -0.21665953,
        -0.03542945,  0.29156857],
       [-1.21564337,  0.09005041,  0.18133163, ..., -0.21665953,
        -1.65522476,  0.29156857],
       [ 0.98373415,  0.77689313,  0.18133163, ..., -0.21665953,
        -0.03542945,  0.29156857]])

In [19]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=1)

# KNN Classifier

In [20]:
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from sklearn.metrics import classification_report
knn = KNeighborsClassifier(n_neighbors=25, metric='euclidean')
knn.fit(x_train, y_train)

KNeighborsClassifier(metric='euclidean', n_neighbors=25)

In [21]:
y_pred = knn.predict(x_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.8423584809089979

# Decision Tree Classifier as Entropy

In [22]:
from sklearn.tree import DecisionTreeClassifier as dtree
ct = dtree(
           criterion="entropy",    # Alternative 'gini'
           max_depth=None       # Alternative, specify an integer
                              # 'None' means full tree till single leaf
           )
ct.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy')

In [23]:
y_te = ct.predict(x_test)

In [24]:
accuracy_score(y_test,y_te)

0.8231139318251612

# Decision Tree Classifier as Gini

In [25]:
from sklearn.tree import DecisionTreeClassifier as dtree
ct = dtree(
           criterion="gini",   
           max_depth=None       # Alternative, specify an integer
                              # 'None' means full tree till single leaf
           )
ct.fit(x_train,y_train)

DecisionTreeClassifier()

In [26]:
y_tes = ct.predict(x_test)

In [27]:
accuracy_score(y_test,y_tes)

0.8262872351315386

# Random Forest Classifier

In [28]:
from sklearn.ensemble import RandomForestClassifier
clf=RandomForestClassifier(n_estimators=100)
clf.fit(x_train,y_train)

y_pred1=clf.predict(x_test)
accuracy_score(y_test,y_pred1)

0.8544375063977889

# KNN Accuracy : 0.8423584809089979

# DTree Gini Accuracy : 0.8231139318251612

# DTree Entropy Accuracy: 0.8262872351315386

# Random Forest Accuracy: 0.8544375063977889

# Through the 4 accuracy tests we can see that the highest accuracy is using Random Forest, i.e., 85.44%